feat(core): introduce local scope & make NgModules optional in ivy

mgechev opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

The templates of a component are compiled based on the NgModule transitive scope, without accepting local component scope.

Issue Number: N/A


## What is the new behavior?

The templates of the component could be only compiled using the local component scope, which makes NgModules optional.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -236,6 +236,14 @@ export class ComponentDecoratorHandler implements
+      const expr = component.get('deps') !;
+      const declarationMeta = staticallyResolve(expr, this.reflector, this.checker);
+      localScope = resolveTypeList(expr, declarationMeta, 'deps', this.reflector)
+                       .map(ref => new WrappedNodeExpr(ref));
```
`WrappedNodeExpr` represents a TypeScript `ts.Node` as an `Expression`. This allows TS nodes to be passed directly through the `@angular/compiler` APIs (which know nothing about TS) and appear in generated expressions.

`ref` in this case is a `Reference<ts.Declaration>` which is the correct way to refer to a node in AOT. `Reference` supports `.toExpression()` to construct an `Expression` that will evaluate to the referenced type within the context of a particular source file (which might decide to use a `WrappedNodeExpr` of the type's `ts.Identifier` if appropriate).

I think `localScope` here should probably be of type `Reference<ts.Declaration>[]` instead of `Expression[]`.

mgechev replied:
Done

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/selector_scope.ts:
```diff
@@ -131,21 +131,22 @@ export class SelectorScopeRegistry {
   }
 
-  lookupCompilationScopeAsRefs(node: ts.Declaration): CompilationScope<Reference>|null {
+  lookupCompilationScopeAsRefs(node: ts.Declaration, localScope: Expression[]|null):
```
I don't think this function should have any knowledge of `localScope`, but it does seem to be unavoidable with the current design.

I'll make a note to do some future refactoring here, but for now this should be fine (except the type should be `Reference<ts.Declaration>[]` as mentioned).

mgechev replied:
What's the refactoring you're thinking of?

mgechev replied:
Updated the implementation so that we don't have `localScope`.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/selector_scope.ts:
```diff
@@ -159,7 +160,12 @@ export class SelectorScopeRegistry {
+    let compilation: Reference<ts.Declaration>[] = [];
+    if (module) {
+      compilation =
+          this.lookupNgModuleScopesOrDie(module !, /* ngModuleImportedFrom */ null).compilation;
```
There should be no need for the `!` - TypeScript knows it can't be null/undefined here.

Also, prefer explicit conditionals: `module !== null` instead of `module`

mgechev replied:
Done

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/test/component_spec.ts:
```diff
@@ -13,6 +13,8 @@ import {TypeScriptReflectionHost} from '../../metadata';
 import {getDeclaration, makeProgram} from '../../testing/in_memory_typescript';
 import {ResourceLoader} from '../src/api';
 import {ComponentDecoratorHandler} from '../src/component';
+import {DirectiveDecoratorHandler} from '../src/directive';
```
Make separate spec files for `DirectiveDecoratorHandler` and `PipeDecoratorHandler`.

mgechev replied:
These imports were redundant, removed them.

alxhub commented on packages/core/src/metadata/directives.ts:
```diff
@@ -552,6 +552,11 @@ export interface Component extends Directive {
+  /**
+   * A set of directives and pipes used in the component's template.
+   */
+  deps?: Array<Type<any>>;
```
What happens if this is used in applications compiled with View Engine? This shouldn't be allowed, and we should give an informative error message if someone tries.

alxhub commented on packages/core/src/render3/jit/directive.ts:
```diff
@@ -71,8 +72,19 @@ export function compileComponent(type: Type<any>, metadata: Component): void {
         // allows the component to patch itself with directiveDefs from the module after it
         // finishes compiling.
-        if (hasSelectorScope(type)) {
+        if (hasSelectorScope(type) && metadata.deps) {
```
Prefer explicit conditionals (`metadata.deps !== undefined`)

mgechev replied:
Done

alxhub commented on packages/core/src/render3/jit/directive.ts:
```diff
@@ -188,3 +200,18 @@ function isViewQuery(value: any): value is Query {
   return value.split(',').map(piece => piece.trim());
 }
+
+function localScopeFor(deps: Array<Type<any>>): NgModuleTransitiveScopes {
```
Since local scopes are just a set of directives and pipes to use in a single component and aren't available in a transitive context, there's really no reason this function should return `NgModuleTransitiveScopes`. Merely returning a set of directives and pipes is fine.

alxhub commented on packages/core/src/render3/jit/directive.ts:
```diff
@@ -71,8 +72,19 @@ export function compileComponent(type: Type<any>, metadata: Component): void {
-        if (hasSelectorScope(type)) {
+        if (hasSelectorScope(type) && metadata.deps) {
           const scopes = transitiveScopesFor(type.ngSelectorScope);
+          const localScope = localScopeFor(metadata.deps);
```
Can we restructure the logic here so we construct the overall scope first, and then call `patchComponentDefWithScope` unconditionally at the end?

mgechev replied:
I updated the logic so we don't call `patchComponentDefWithScope` in each branch of the conditional statements. We still need to invoke it only when there are directives/pipes, otherwise `directiveDefs` and `pipeDefs` should be `null`.

mgechev replied:
Removed the `localScope` and moved the `deps`.

alxhub commented on packages/core/test/render3/ivy/BUILD.bazel:
```diff
@@ -23,5 +23,6 @@ jasmine_node_test(
     deps = [
         ":ivy_lib",
         "//packages/core/test/render3:domino",
+        "//packages/core/test/render3:render3_lib",
```
We should break this target up so we're not pulling in the specs from `//packages/core/test/render3:render3_lib` as well. Maybe move it into `//packages/core/test/render3:util`?

alxhub commented on packages/compiler/src/render3/view/api.ts:
```diff
@@ -125,6 +125,11 @@ export interface R3ComponentMetadata extends R3DirectiveMetadata {
+  /**
+   * Dependencies of the component's template which could be pipes or directives.
+   */
+  localScope: o.Expression[]|null;
```
I don't think this is needed anymore - we should be including the local dependencies into the `directives` and `pipes` sections of `R3ComponentMetadata`.

mgechev replied:
I introduced `localScope` because during the `analyze` call we don't have metadata for the individual symbols and thus we'll break locality if we try to resolve it. Without the metadata we don't know if the symbols are directives or pipes.

mgechev commented:
The proposed PR structure makes sense to me and it should be part of the documentation (probably [DEVELOPER.md](https://github.com/angular/angular/blob/master/docs/DEVELOPER.md)?) We can discuss the right place for it in a different channel.

mary-poppins commented:
You can preview 8a00459 at https://pr27481-8a00459.ngbuilds.io/.

mary-poppins commented:
You can preview 21bf4d4 at https://pr27481-21bf4d4.ngbuilds.io/.

mary-poppins commented:
You can preview dbc1934 at https://pr27481-dbc1934.ngbuilds.io/.

JoostK commented on packages/core/test/render3/ivy/jit_spec.ts:
```diff
@@ -377,6 +380,32 @@ ivyEnabled && describe('render3 jit', () => {
+    const def = (TestComponent as any).ngComponentDef;
+    expect(def).not.toBeNull();
+    const myComp = renderComponent(TestComponent as any);
+    const result = getHostElement(myComp).innerHTML;
```
This logic is available as `ComponentFixture` (the Ivy one for tests, not the one for `TestBed`).

JoostK reviewed (commented):
Awesome! 🎉

robwormald commented on packages/core/src/metadata/directives.ts:
```diff
@@ -552,6 +552,11 @@ export interface Component extends Directive {
+  /**
+   * A set of directives and pipes used in the component's template.
+   */
+  deps?: Array<Type<any>>;
```
Should this take `Array<Type<any>> | Array<Array<Type<any>>>` and get smooshed, er, flatten'd? 

Probably the broader question is that if we expect developers to be lazy, as developers are, presumably we'll see something like pre-RC5 usage:

```ts
//lib.ts
export class MyDirectiveA {}
export class MyDirectiveB {}

export const MY_DIRECTIVES = [
  MyDirectiveA, 
  MyDirectiveB
];
```

```ts
import {MY_DIRECTIVES} from './lib.ts'

@Directive()
export class AppDirective {}

@Component({
  deps: [AppDirective, ...MY_DIRECTIVES] //should users be responsible for spreading?
})
export class MyApp {}
```
This has implications on treeshaking of course.

robwormald replied:
throwback to old idea: https://github.com/angular/angular-cli/issues/96#issuecomment-163528438

could potentially be optimized as a prod step?, which seems to be tree-shaking friendly: https://bit.ly/2PlrnSM

robwormald replied:
All components are also directives, but that's something a new user has to learn out-of-band - mentioning components  in the jsdoc explicitly here might be helpful?

mgechev replied:
Added components as well. For the `deps` type, I'd go with `Array<Type<any>>`. IMO, it makes the API more obvious.

@mhevery may have an opinion on nested arrays.

manekinekko replied:
I do backup Rob's comment as I see that practice a lot in userland. Also, many developers will expect the `deps` property to work the same as the `providers` property.

mgechev replied:
Looking at the pre-NgModules [releases](https://github.com/mgechev/angular-seed/blob/2.0.0-rc.4/src/client/app/app.component.ts#L16) (rc.4), the corresponding properties were nested arrays as well. To keep consistent with `providers` and `viewProviders` it makes sense to update the type.

mgechev replied:
At first it also makes sense to annotate `deps` with `@internal` until we decide how/if we want to ship this.

mary-poppins commented:
You can preview 7ff05f1 at https://pr27481-7ff05f1.ngbuilds.io/.

mary-poppins commented:
You can preview ef7624d at https://pr27481-ef7624d.ngbuilds.io/.

mgechev committed:
refactor: move resolveTypeList to utils

Move the `resolveTypeList` function to `utils` so it can be reused in
`Component` and `NgModule` decorator handlers.

mgechev committed:
feat(core): add deps property to component

Introduce `deps` property in the configuration object passed to the `Component` decorator. The `deps` property contains a list of providers and directives which are used in the component's template. This makes components independent from NgModules and simplifies the API for `@angular/elements` and `@angular/core`.

mary-poppins commented:
You can preview 8a8fd93 at https://pr27481-8a8fd93.ngbuilds.io/.

mary-poppins commented:
You can preview 285218c at https://pr27481-285218c.ngbuilds.io/.

mary-poppins commented:
You can preview b03f1c5 at https://pr27481-b03f1c5.ngbuilds.io/.

mary-poppins commented:
You can preview 24a547d at https://pr27481-24a547d.ngbuilds.io/.

mary-poppins commented:
You can preview 9d78ce2 at https://pr27481-9d78ce2.ngbuilds.io/.

mary-poppins commented:
You can preview 3fba7e7 at https://pr27481-3fba7e7.ngbuilds.io/.

mary-poppins commented:
You can preview 2192263 at https://pr27481-2192263.ngbuilds.io/.

mary-poppins commented:
You can preview 56d852e at https://pr27481-56d852e.ngbuilds.io/.

mary-poppins commented:
You can preview f10045d at https://pr27481-f10045d.ngbuilds.io/.

mary-poppins commented:
You can preview 348cc92 at https://pr27481-348cc92.ngbuilds.io/.

mary-poppins commented:
You can preview f7e7985 at https://pr27481-f7e7985.ngbuilds.io/.

trotyl commented on packages/core/src/render3/jit/directive.ts:
```diff
@@ -196,3 +212,9 @@ function isViewQuery(value: any): value is Query {
+function localScopeFor(deps: Array<Type<any>>) {
+  const pipes = new Set(deps.filter(dep => !!getPipeDef(dep)));
+  const directives = new Set(deps.filter(dep => !!getDirectiveDef(dep) || !!getComponentDef(dep)));
+  return {directives, pipes};
```
It seems that JIT implementation supports multi-purpose class while AOT doesn't?

```typescript
@Directive({})
@Pipe({})
class MyMultiPurposeClass implements PipeTransform { }
```

mgechev replied:
Shouldn't be. Updated the implementation.

mary-poppins commented:
You can preview a1966ac at https://pr27481-a1966ac.ngbuilds.io/.

mgechev commented:
Closing the PR for now; we can reopen it once we land more critical blocking work.
